Lacrit/assignment1/q2.cpp

```diff
@@ -4,44 +4,76 @@
 template<class T> 
 struct Node
 {
-    T i;
-    struct Node *next;
+	private:
+    	T i;
+		Node *next;
+	public: 
+		Node* get_next()
+		{
+			return next;
+		}  
+		T get_i()
+		{
+			return i;
+		}
+		void set_next(Node *new_next)
+		{
+			next = new_next;
+		}
+		void set_i(T j)
+		{
+			i = j;
+		}
+			
 };
 
 
 template<class T> 
 T findNode (Node<T> *listHead, int n)
 {
+	 if (listHead == NULL) return 0; 
      Node<T> *ptr1, *ptr2;  // we need 2 pointers
      ptr1 = ptr2 = listHead; // set the pointers to point to the list head initially
 
-    while(ptr1->next != NULL) // keep looping until we reach the tail (next will be NULL for the last node)
+    while(ptr1->get_next() != NULL) // keep looping until we reach the tail (next will be NULL for the last node)
     {
         if(n > 0)
         {
-            ptr1 = ptr1->next; //increment only the 1st pointer
+            ptr1 = ptr1->get_next(); //increment only the 1st pointer
             n--;
         }
         else
         {
-            ptr1 = ptr1->next; //increment both pointers
-            ptr2 = ptr2->next;
+            ptr1 = ptr1->get_next(); //increment both pointers
+            ptr2 = ptr2->get_next();
         }
    }
-   return ptr2->i;
+   return ptr2->get_i();
 }
 
 template<class T> 
 void push(Node<T>** head, T data)
 {
-  Node<T>* new_node =(Node<T>*) malloc(sizeof(Node<T>));
- 
-  new_node->i  = data;
-  new_node->next = (*head);
+  //Node<T>* new_node =(Node<T>*) malloc(sizeof(Node<T>))
+  //Just left the upper row for reference. Correction:
+  Node<T>* new_node = new Node<T>(); // no need to pass the size
```

Comment:
If you add a constructor to Node, you can replace this 3 lines with:
`Node<T>* new_node = new Node<T>(data, *head);`
Or even the whole function with:
`head = new Node<T>(data, *head);`
:)